ccmbioinfo/lfs
# CARDS-2134: Fix performance metrics generated by Cards4PREMs

This PR implements CARDS-2134 and, therefore, when CARDS is started in `cards4prems` mode with a properly configured `SLACK_PERFORMANCE_URL` environment variable, the following performance metrics will be periodically sent to a Slack channel:

- For each of the 5 types of surveys (OED, CPESIC, Rehab, OAIP, OED + OAIP):
  - Number of initial emails sent
  - Number of 1st reminder emails sent
  - Number of 2nd reminder emails sent
- Number of imported appointments
- Number of surveys submitted
- Number of patients who responded

Testing Instructions
--------------------------

1. Build this (`CARDS-2134`) branch (including a _development_ Docker image) with `mvn clean install -Pdocker`.
2. Create the `~/cards-2134.env` file and fill it with the following contents
```
NIGHTLY_NOTIFICATIONS_SCHEDULE=0 * * ? * *
NIGHTLY_SLACK_NOTIFICATIONS_SCHEDULE=30 * * ? * *
SLACK_PERFORMANCE_URL={{PLEASE DIRECT MESSAGE ME ON SLACK FOR THIS}}
```
3. `cd compose-cluster`
4. `mkdir ~/test_mail`
5. `python3 generate_compose_yaml.py --dev_docker_image --oak_filesystem --composum --adminer --cards_project cards4prems --mssql --smtps --smtps_test_container --smtps_test_mail_path ~/test_mail --custom_env_file ~/cards-2134.env`
6. Ensure that you have built the [`cards/postfix-docker` image](https://github.com/data-team-uhn/cards/blob/dev/Utilities/Development/EmailServer/README.md).
7. `docker-compose build`
8. `docker-compose up -d`
9. Visit http://localhost:1435, and login to the MS-SQL database as follows:
    - **System**: _MS SQL (beta)_
    - **Server**: `mssql`
    - **Username**: `sa`
    - **Password**: `testPassword_`
    - **Database**: `master`
11. Import the `compose-cluster/mssql/cards-2134-test.sql` file into the MS-SQL database.
12. Visit http://localhost:8080 and login as `admin`:`admin`.
13. Visit http://localhost:8080/system/console/components and disable the `io.uhndata.cards.prems.internal.importer.UpdatedDischargeDateFiller` component.
14. Visit http://localhost:8080/system/console/configMgr and edit the `Clarity import filter - CPESIC percentage` configuration settings the values as follows:
    - **Default Frequency**: `0`
    - **Per Department Frequency**: `TW-3B Fell Pavilion = 1`
15. Import all the _mock_ Clarity data into CARDS by loading the following URLs:
    - http://localhost:8080/Subjects.importClarity?pastDayToQuery=5
    - http://localhost:8080/Subjects.importClarity?pastDayToQuery=7
    - http://localhost:8080/Subjects.importClarity?pastDayToQuery=9
    - http://localhost:8080/Subjects.importClarity?pastDayToQuery=11
16. Reload the CARDS dashboard. A total of 28 _Patients_, each with one _Visit_ (therefore 28 _Visits_) should be created.
17. Click the **Send / Receive** button in _GNOME Evolution_. You should be seeing emails for all past _Visits_ **except** for those which were _5_ days in the past (`21@mail.com`, `22@mail.com` ... `27@mail.com`).
18. Complete some of the surveys as if you were a patient (note: you'll need to change `https://localhost:8080` to `http://localhost:8080` in the link).
19. Check the Slack channel displaying these performance metric updates. Ensure that all _Total_ metrics have a positive value.

![Screenshot_2023-03-06_12-23-59](https://user-images.githubusercontent.com/5349728/223185382-5ba105fb-9071-4188-9e0c-d142ca9e8677.png)
